src/ray/core_worker/task_event_buffer.cc

```diff
@@ -0,0 +1,193 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/task_event_buffer.h"
+
+namespace ray {
+namespace core {
+
+namespace worker {
+
+TaskEventBufferImpl::TaskEventBufferImpl(std::unique_ptr<gcs::GcsClient> gcs_client)
+    : work_guard_(boost::asio::make_work_guard(io_service_)),
+      periodical_runner_(io_service_),
+      gcs_client_(std::move(gcs_client)) {}
+
+Status TaskEventBufferImpl::Start(bool auto_flush) {
+  auto report_interval_ms = RayConfig::instance().task_events_report_interval_ms();
+  RAY_CHECK(report_interval_ms > 0)
+      << "RAY_task_events_report_interval_ms should be > 0 to use TaskEventBuffer.";
+
+  buffer_.reserve(RayConfig::instance().task_events_max_num_task_events_in_buffer());
+
+  io_thread_ = std::thread([this]() {
+#ifndef _WIN32
+    // Block SIGINT and SIGTERM so they will be handled by the main thread.
```

Comment:
nice! Just curious if we block the signal, is it going into the main thread that has a handler or is it just simply ignored?

Reply:
I think the coreworker will handle it instead.